### PR TITLE
Switch to sbt-datatype, sjson-new & Vectors

### PR DESCRIPTION
**Pull request by dwijnand**
Review please by @eed3si9n.

### CODE REVIEW

**Comment by dwijnand** [comment]
Or let me know how you think I should break this down and/or simplify it to ease reviewing it.

File: librarymanagement/src/main/scala/sbt/librarymanagement/Artifact.scala [hunk 1]
```diff
@@ -5,30 +5,47 @@ package sbt.librarymanagement
 
 import java.io.File
 import java.net.URL
-import sbt.serialization._
 
-final case class Artifact(name: String, `type`: String, extension: String, classifier: Option[String], configurations: Iterable[Configuration], url: Option[URL], extraAttributes: Map[String, String]) {
+abstract class ArtifactParent {
```
Comment (eed3si9n) [inline librarymanagement/src/main/scala/sbt/librarymanagement/Artifact.scala]: Is this the mixin for the methods? If so the naming convention I think should be `ArtifactExtra`.
Sort of like `ProjectExtra` etc - https://github.com/sbt/sbt/blob/0.13/main/src/main/scala/sbt/Project.scala#L608

**Review by eed3si9n — CHANGES_REQUESTED** [review]
Minor comment, but mostly looks good.

File: librarymanagement/src/main/scala/sbt/internal/librarymanagement/formats/Function1Format.scala [hunk 2]
```diff
@@ -0,0 +1,8 @@
+package sbt.internal.librarymanagement.formats
+
+import sjsonnew._
+
+trait Function1Format { self: BasicJsonProtocol =>
+  implicit def Function1Format[T, U]: JsonFormat[T => U] =
+    project(MyCrazyReferences.referenced _, (ref: String) => MyCrazyReferences(ref, classOf[T => U]))
```
Comment (dwijnand) [inline librarymanagement/src/main/scala/sbt/internal/librarymanagement/formats/Function1Format.scala]: @eed3si9n How do you feel about this?

File: librarymanagement/src/main/scala/sbt/librarymanagement/Artifact.scala [hunk 3]
```diff
@@ -5,30 +5,47 @@ package sbt.librarymanagement
...
 
-object Artifact {
-  def apply(name: String): Artifact = Artifact(name, DefaultType, DefaultExtension, None, Nil, None)
-  def apply(name: String, extra: Map[String, String]): Artifact = Artifact(name, DefaultType, DefaultExtension, None, Nil, None, extra)
-  def apply(name: String, classifier: String): Artifact = Artifact(name, DefaultType, DefaultExtension, Some(classifier), Nil, None)
-  def apply(name: String, `type`: String, extension: String): Artifact = Artifact(name, `type`, extension, None, Nil, None)
-  def apply(name: String, `type`: String, extension: String, classifier: String): Artifact = Artifact(name, `type`, extension, Some(classifier), Nil, None)
-  def apply(name: String, url: URL): Artifact = Artifact(name, extract(url, DefaultType), extract(url, DefaultExtension), None, Nil, Some(url))
-  def apply(name: String, `type`: String, extension: String, classifier: Option[String], configurations: Iterable[Configuration], url: Option[URL]): Artifact =
+abstract class ArtifactCompanion {
```
Comment (dwijnand) [inline librarymanagement/src/main/scala/sbt/librarymanagement/Artifact.scala]: @eed3si9n What about the companion? Should it be `ArtifactExtra` too?

Comment (eed3si9n) [reply librarymanagement/src/main/scala/sbt/librarymanagement/Artifact.scala]: `ArtifactFunctions`, maybe.

File: librarymanagement/src/main/scala/sbt/internal/librarymanagement/formats/Function1Format.scala (hunk 2 above)
Comment (eed3si9n) [reply librarymanagement/src/main/scala/sbt/internal/librarymanagement/formats/Function1Format.scala]: The name looks crazy. Anything containing the function itself probably should be handled as reference pattern, not the function.

**Review by eed3si9n — APPROVED** [review]

**Pull request merged**